### PR TITLE
support sign with hdwallet account

### PR DESCRIPTION
**Pull request by millken**
add prefix `hdw::` for signer '-s' parameter

### CODE REVIEW

File: ioctl/cmd/action/action.go [hunk 1]
```diff
@@ -108,9 +110,40 @@ func decodeBytecode() ([]byte, error) {
...
+			return 0, 0, 0, output.NewError(output.InputError, fmt.Sprintf("%v must be integer value", args[i]), err)
+		}
+		arg[i] = uint32(u64)
+	}
+	return arg[0], arg[1], arg[2], nil
+}
+
+func aliasIsHdwalletKey(addressOrAlias string) bool {
+	if strings.HasPrefix(strings.ToLower(addressOrAlias), "hdw::") {
+		return true
```
Comment (dustinxie) [inline ioctl/cmd/action/action.go:135]: should return whether the hdwallet config file exists

File: ioctl/cmd/action/action.go [hunk 2]
```diff
@@ -108,9 +110,40 @@ func decodeBytecode() ([]byte, error) {
 	return hex.DecodeString(util.TrimHexPrefix(bytecodeFlag.Value().(string)))
 }
 
+func parseHdwPath(addressOrAlias string) (uint32, uint32, uint32, error) {
```
Comment (dustinxie) [inline ioctl/cmd/action/action.go:113]: add test for this func

File: ioctl/cmd/hdwallet/hdwalletcreate.go [hunk 3]
```diff
@@ -7,11 +7,15 @@
 package hdwallet
 
 import (
+	"bytes"
 	"fmt"
 	"io/ioutil"
 
+	ecrypt "github.com/ethereum/go-ethereum/crypto"
+	"github.com/iotexproject/go-pkgs/crypto"
```
Comment (dustinxie) [inline ioctl/cmd/hdwallet/hdwalletcreate.go:15]: this is internal package, move to bottom
and run 'make fmt'

File: ioctl/cmd/hdwallet/hdwalletcreate.go [hunk 4]
```diff
@@ -82,6 +85,62 @@ func hdwalletCreate() error {
 
 	output.PrintResult(fmt.Sprintf("Mnemonic pharse: %s\n"+
 		"It is used to recover your wallet in case you forgot the password. Write them down and store it in a safe place.", mnemonic))
-
 	return nil
 }
+
+// DeriveKey derives the key according to path
+func DeriveKey(account, change, index uint32, password string) (string, crypto.PrivateKey, error) {
+	// derive key as "m/44'/304'/account'/change/index"
```
Comment (dustinxie) [inline ioctl/cmd/hdwallet/hdwalletcreate.go:93]: i think current test case only cover path /0/a/b
add test case for path /a/b/c

File: ioctl/cmd/hdwallet/hdwalletcreate.go [hunk 5]
```diff
@@ -7,11 +7,15 @@
...
+	"bytes"
 	"fmt"
 	"io/ioutil"
 
+	ecrypt "github.com/ethereum/go-ethereum/crypto"
+	"github.com/iotexproject/go-pkgs/crypto"
+	"github.com/iotexproject/iotex-address/address"
+	hdwallet "github.com/miguelmota/go-ethereum-hdwallet"
 	"github.com/spf13/cobra"
-
```
Comment (dustinxie) [inline ioctl/cmd/hdwallet/hdwalletcreate.go:14]: remove empty line

File: ioctl/cmd/action/action.go (hunk 2 above)
Comment (millken) [reply ioctl/cmd/action/action.go:113]: see DeriveKey

File: ioctl/cmd/action/action.go (hunk 1 above)
Comment (millken) [reply ioctl/cmd/action/action.go:135]: see DeriveKey

File: ioctl/util/util_test.go [hunk 6]
```diff
@@ -64,3 +64,31 @@ func TestTrimHexPrefix(t *testing.T) {
...
+func TestParseHdwPath(t *testing.T) {
+	r := require.New(t)
+
+	tests := []struct {
+		addressOrAlias string
+		a, b, c        uint32
+		err            string
+	}{
+		{"hdw::0/1/2", 0, 1, 2, ""},
+		{"hdw::0/1", 0, 1, 0, ""},
```
Comment (dustinxie) [inline ioctl/util/util_test.go:77]: great, this is a good comprehensive test suite

hdw::0/1 should return 0, 0, 1, when there are 2 values they should be treated as 0/b/c

File: ioctl/cmd/hdwallet/hdwallet_test.go [hunk 7]
```diff
@@ -91,6 +91,26 @@ func Test_Hdwallet(t *testing.T) {
...
+	private, err = wallet.PrivateKey(account)
+	require.NoError(err)
+	addr3, err := address.FromBytes(hashECDSAPublicKey(&private.PublicKey))
+	require.NoError(err)
+
+	require.Equal(addr2, addr3)
+
+	account1 = 123
+	derivationPath = fmt.Sprintf("m/44'/304'/%d'/%d/%d", account1, change, index)
+	require.Equal("m/44'/304'/123'/1/2", derivationPath)
```
Comment (dustinxie) [inline ioctl/cmd/hdwallet/hdwallet_test.go:113]: derive a key using /123/1/2, and verify it != addr2

File: ioctl/cmd/account/accountsign.go [hunk 8]
```diff
@@ -51,9 +51,15 @@ func init() {
...
-		return output.NewError(output.InputError, "failed to get signer addr", err)
+	var addr string
+	var err error
+	if !util.AliasIsHdwalletKey(signer) {
+		addr, err = util.GetAddress(signer)
+		if err != nil {
+			return output.NewError(output.InputError, "failed to get signer addr", err)
+		}
+	} else {
+		addr = signer
```
Comment (dustinxie) [inline ioctl/cmd/account/accountsign.go:62]: can get rid of else {}
```go
addr := signer
if !util.AliasIsHdwalletKey(signer) {
        var err error
        addr, err = util.GetAddress(signer)
	if err != nil {
		return output.NewError(output.InputError, "failed to get signer addr", err)
	}
} 
```

File: ioctl/cmd/hdwallet/hdwalletcreate.go [hunk 9]
```diff
@@ -82,6 +85,62 @@ func hdwalletCreate() error {
...
+	if !bytes.Equal(hash, util.HashSHA256(mnemonic)) {
+		return "", nil, output.NewError(output.ValidationError, "password error", nil)
+	}
+
+	wallet, err := hdwallet.NewFromMnemonic(string(mnemonic))
+	if err != nil {
+		return "", nil, err
+	}
+
+	derivationPath := fmt.Sprintf("m/44'/304'/%d'/%d/%d", account, change, index)
```
Comment (dustinxie) [inline ioctl/cmd/hdwallet/hdwalletcreate.go:125]: should use `DefaultRootDerivationPath` here, also i think it should be easier to define:
```go
const DefaultRootDerivationPath = "m/44'/304'"
```
and Sprintf "/a/b/c" after it

File: ioctl/cmd/hdwallet/hdwallet_test.go [hunk 10]
```diff
@@ -91,6 +91,26 @@ func Test_Hdwallet(t *testing.T) {
 
 	require.Equal(addr1, addr2)
 
+	//test DeriveKey
+	account1 := 0
+	change := 1
+	index := 2
+
+	derivationPath = fmt.Sprintf("m/44'/304'/%d'/%d/%d", account1, change, index)
```
Comment (dustinxie) [inline ioctl/cmd/hdwallet/hdwallet_test.go:99]: use `DefaultRootDerivationPath`

**Review by dustinxie — APPROVED** [review]

**Pull request merged**